### DISCUSSION TITLE
etcd client retry behavior

### DISCUSSION THREAD

**Discussion by marriner-stripe** [etcd-io/etcd#14791] [Q&A]
I have a few questions around retry behavior.  We are planning on using etcd in scenarios where there are many clients interacting with a single cluster.

We are worried about the default retry behavior on Gets([100 retries](https://github.com/etcd-io/etcd/blob/main/client/v3/options.go#L45) on Service Unavailable errors) causing self-sustaining overloads and are considering disabling/changing retries. 

1. Is there any way to change the retry policy for the client other than disabling retries entirely? Even hacky approaches would work for us for now! The best approach we were able to find was to change the client to the non-retrying version, e.g. 

```
etcdClient.KV = etcdv3.NewKVFromKVClient(etcdpb.NewKVClient(etcdClient.ActiveConnection()), etcdClient)
```

2. Are there any known downsides to disabling retries, other than lower availability on a flaky network?  Were the values of this specific retry policy chosen intentionally or are they generic?

**Comment by lavacat** [comment]
I don't think there is a way to change policy. Looking at git blame, seams like @ptabor updated TODO to use [RetryPolicy](https://github.com/etcd-io/etcd/pull/12706/files#diff-2bbaa3c83234888775aabf7697076e8035451045c479938e03ebb5c929132679R212)
```
	- // TODO: Replace all of clientv3/retry.go with interceptor based retry, or with
	- // https://github.com/grpc/proposal/blob/master/A6-client-retries.md#retry-policy
	- // once it is available.
	+ // TODO: Replace all of clientv3/retry.go with RetryPolicy:
	+ // https://github.com/grpc/grpc-proto/blob/cdd9ed5c3d3f87aef62f373b93361cf7bddc620d/grpc/service_config/service_config.proto#L130
```
That's probably best direction to follow if you want to contribute this change.

Another option is to make `defaultUnaryMaxRetries`, `defaultStreamMaxRetries`, `defaultBackoffWaitBetween`, `defaultBackoffJitterFraction` configurable. Similar issue https://github.com/etcd-io/etcd/issues/13344

**Comment by lavacat** [comment]
> We are worried about the default retry behavior on Gets([100 retries](https://github.com/etcd-io/etcd/blob/main/client/v3/options.go#L45) on Service Unavailable errors) causing self-sustaining overloads and are considering disabling/changing retries.

Have you done load testing for your use case? What types of [Unavailable](https://github.com/etcd-io/etcd/blob/e2d67f2e3bfa6f72178e26557bb22cc1482c418c/api/v3rpc/rpctypes/error.go#L73) errors are you seeing? And what types of `repeatable` requests generate them?
